pkg/kube/client/ssa/caching_unstructured_extractor.go

```diff
@@ -0,0 +1,290 @@
+package ssa
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/managedfields"
+	applymetav1 "k8s.io/client-go/applyconfigurations/meta/v1"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/kube-openapi/pkg/handler3"
+	"k8s.io/kube-openapi/pkg/schemaconv"
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+	smdschema "sigs.k8s.io/structured-merge-diff/v4/schema"
+	"sigs.k8s.io/structured-merge-diff/v4/typed"
+)
+
+// cachingUnstructuredExtractor is a caching implementation of v1.UnstructuredExtractor
+// using OpenAPI V3 discovery information.
+type cachingUnstructuredExtractor struct {
+	cache *GvkParserCache
+	dc    discovery.DiscoveryInterface
+}
+
+// NewCachingUnstructuredExtractor returns a new cachingUnstructuredExtractor
+func NewCachingUnstructuredExtractor(dc discovery.DiscoveryInterface, cache *GvkParserCache) (applymetav1.UnstructuredExtractor, error) {
+	return &cachingUnstructuredExtractor{
+		dc:    dc,
+		cache: cache,
+	}, nil
+}
+
+// Extract extracts the applied configuration owned by fieldManager from an unstructured object.
+// Note that the apply configuration itself is also an unstructured object.
+func (e *cachingUnstructuredExtractor) Extract(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "")
+}
+
+// ExtractStatus is the same as ExtractUnstructured except
+// that it extracts the status subresource applied configuration.
+// Experimental!
+func (e *cachingUnstructuredExtractor) ExtractStatus(object *unstructured.Unstructured, fieldManager string) (*unstructured.Unstructured, error) {
+	return e.extractUnstructured(object, fieldManager, "status")
+}
+
+// schemerForGVK queries the openapi v3 discovery endpoint
+// and builds a custom openapi.GroupVersion to retrieve the schema
+// only for the GroupVersion of the given GroupVersionKind
+// It also stores the returned ETag associated with the GV for caching
+func (e *cachingUnstructuredExtractor) schemerForGVK(gvk schema.GroupVersionKind) (*CustomOAPIGroupVersion, error) {
```

Comment:
making this an interface which has got one `Schema(ctx, contentType)` and another `ETag() string` would make it a little bit easier to see what kind of interface you're expecting. I've never heard the term `schemer` before in this context, what does that mean?

Reply:
done